Emit any pre-connnect auth errors in a way the onConnect listeners will handle

SimonWoolf opened the pull request:
Fixes https://github.com/ably/ably-js/issues/311

@paddybyers would you mind explaining a couple things when you have a sec?

- What's the semantic difference between finishing a transport with `failed` (which is what aborting does) and finishing it with `error`? Any reason not to merge the two events, and avoid bugs like this where something's listening for one but not the other?

- You have a `/* FIXME: this should not be fatal */` in the websocket `onerror` handler. Why not? Is it possible it'll ever be called with a nonfatal error? (the w3 spec is [very vague](https://www.w3.org/TR/websockets/#handler-websocket-onerror), and the websocket spec isn't much help either - it defines a lot of codes for the close frame, but doesn't mention any nonfatal errors that I can find). And if nonfatal errors are possible, why have the abort there at all -- why not just log them and rely on the `onclose` being called immediately after if the error was fatal?

mattheworiordan commented:
Nice work @SimonWoolf

paddybyers commented:
> Any reason not to merge the two events, and avoid bugs like this where something's listening for one but not the other?

No, I don't think so. I might have had something in mind at the time, but it was a long time ago :)

> Is it possible it'll ever be called with a nonfatal error?

Again, I was obviously thinking of something at the time, but now I can't remember. Maybe some error conditions such as an individual frame exceeding maximum length or something like that needn't necessarily mean the transport needs to die. But in our protocol it would mean that, so I can't think of any reason now to make it non-fatal.

paddybyers commented:
LGTM BTW

SimonWoolf commented:
Replaced previous commit with 598fc8d which switches everything to `failed`, per discussion above.

@paddybyers done some more experimenting, and as far as I can tell, onerror is: almost always useless, and if the websocket was connected, always followed by an onclose which contains the actual reason for the error (eg in the exceeded-max-frame-size case it'd be 1009). Aborting the transport in an onerror was if anything harmful, as that disposed of it before the onclose event fires, so we don't find out what the ws close code was. Added de3de72, which puts off the abort into a nexttick (so if it was followed by an onClose, it'll already have been disposed on the onClose's terms by then), and changed onClose to fail the transport if the close was unclean (to match the current behaviour).

paddybyers commented:
ok, LGTM, thanks
